text/0000-network-update-loop.md

```diff
@@ -0,0 +1,706 @@
+    Update = 0,
+    PreLateUpdate = 1,
+    PostLateUpdate = 2
+}
+
+public static class NetworkUpdateLoop
+{
+    [RuntimeInitializeOnLoadMethod]
+    private static void Initialize()
+    {
+        var customPlayerLoop = PlayerLoop.GetCurrentPlayerLoop();
+
+        for (int i = 0; i < customPlayerLoop.subSystemList.Length; i++)
+        {
+            var playerLoopSystem = customPlayerLoop.subSystemList[i];
+
+            if (playerLoopSystem.type == typeof(Initialization))
+            {
+                var subsystems = playerLoopSystem.subSystemList.ToList();
+                subsystems.Add(NetworkInitialization.CreateLoopSystem());
+                playerLoopSystem.subSystemList = subsystems.ToArray();
+            }
+            else if (playerLoopSystem.type == typeof(EarlyUpdate))
+            {
+                var subsystems = playerLoopSystem.subSystemList.ToList();
+                subsystems.Insert(0, NetworkEarlyUpdate.CreateLoopSystem());
+                playerLoopSystem.subSystemList = subsystems.ToArray();
+            }
+            // add/insert other loop systems ...
+
+            customPlayerLoop.subSystemList[i] = playerLoopSystem;
+        }
+
+        PlayerLoop.SetPlayerLoop(customPlayerLoop);
+    }
+
+    private struct NetworkInitialization
+    {
+        public static PlayerLoopSystem CreateLoopSystem()
+        {
+            return new PlayerLoopSystem
+            {
+                type = typeof(NetworkInitialization),
+                updateDelegate = RunNetworkInitialization
+            };
+        }
+    }
+
+    private struct NetworkEarlyUpdate
+    {
+        public static PlayerLoopSystem CreateLoopSystem()
+        {
+            return new PlayerLoopSystem
+            {
+                type = typeof(NetworkEarlyUpdate),
+                updateDelegate = RunNetworkEarlyUpdate
+            };
+        }
+    }
+
+    // define other loop systems ...
+
+    public static NetworkUpdateStage UpdateStage;
+
+    private static readonly List<INetworkUpdateSystem> m_Initialization_List = new List<INetworkUpdateSystem>();
+    private static INetworkUpdateSystem[] m_Initialization_Array = new INetworkUpdateSystem[0];
+
+    private static void RunNetworkInitialization()
+    {
+        UpdateStage = NetworkUpdateStage.Initialization;
+        int arrayLength = m_Initialization_Array.Length;
+        for (int i = 0; i < arrayLength; i++)
+        {
+            m_Initialization_Array[i].NetworkUpdate();
+        }
+    }
+
+    private static readonly List<INetworkUpdateSystem> m_EarlyUpdate_List = new List<INetworkUpdateSystem>();
+    private static INetworkUpdateSystem[] m_EarlyUpdate_Array = new INetworkUpdateSystem[0];
+
+    private static void RunNetworkEarlyUpdate()
+    {
+        UpdateStage = NetworkUpdateStage.EarlyUpdate;
+        int arrayLength = m_EarlyUpdate_Array.Length;
+        for (int i = 0; i < arrayLength; i++)
+        {
+            m_EarlyUpdate_Array[i].NetworkUpdate();
```

Comment:
I know this is pseudocode and the RFC does not specify the underlying implementation, yet thought I'd suggest rather than have individually named variables with and duplicated loop code to invoke it to rather have something like a map of the enums to arrays.  Then you could just have one function for 'Run****' which takes the enum and does the right thing.

Reply:
that is less performant than this approach. I think I will plug some benchmarks into this RFC comparing list, array, hashset, dictionary containers in different places.

Reply:
actually, I'd like to keep these implementation-dependent details to MLAPI implementation PR. I will be comparing, explaining, and discovering trade-offs over there.